lambda/translcore.ml

```diff
@@ -1276,6 +1106,284 @@ and transl_record ~scopes loc env fields repres opt_init_expr =
+      | Array_of_filtered_arrays(arrs) ->
+        let sub_arr = Ident.create_local "sub_arr" in 
+        let sub_arr_len = Ident.create_local "sub_arr_len" in 
+        Llet(Variable, Pintval, i, int 0,
+        Lwhile(Lprim(Pintcomp(Clt), [Lvar(i); arr_len], loc),
+          Lsequence(
+              Llet(Strict, Pgenval,  sub_arr, 
+                Lprim(Parrayrefs(Paddrarray), [arr; Lvar(i)], loc),
+                Llet(Strict, Pintval, sub_arr_len, 
+                  Lprim(Parrayrefs(Paddrarray), 
+                    [arr; Lprim(Paddint, [Lvar(i); int 1], loc)],
+                    loc),
+                  transl_for (Lvar(sub_arr)) (Lvar(sub_arr_len)) arrs)),
+            Lassign(i, Lprim(Paddint, [Lvar(i); int 2], loc))         
+          )))
+    in 
+    (*Remove the outer most layer. The outermost array would always be of size 
+      one, thats why its always unwrapped (arr, arr_len).*)
+    let arrs = 
+      match arrs with
+      | Array_of_elements -> arrs
+      | Array_of_arrays(arrs) -> arrs
+      | Array_of_filtered_arrays(arrs) -> arrs
+    in
+    match res_kind with
+    | Pgenarray ->
+      Llet(Variable, Pgenval, res, empty_arr,
+        Llet(Variable, Pintval, counter, (int 0),  
+          Lsequence(transl_for arr arr_len arrs, Lvar(res))))
+    | Pintarray | Pfloatarray | Paddrarray as kind -> 
+      Llet(Strict, Pgenval, res, make_array_of_kind kind (Lvar(total_len)),
+        Llet(Variable, Pintval, counter, (int 0),  
+          Lsequence(transl_for arr arr_len arrs, Lvar(res))))
+  in
+  (*The global counter is required when the size of the result array is not 
+    known in advance.*)
+  let global_counter = 
+    if (List.length blocks  = 1) && Option.is_none (List.hd blocks).guard
+    then None
+    else Some(Ident.create_local "global_counter") 
+  in
+  let base_block = 
+    transl_block global_counter 
+      (Unguarded(transl_exp ~scopes  body, array_kind), 
+        Array_of_elements) 
+      (List.hd blocks) 
+  in
+  let translated_blocks, arrs = 
+    List.fold_left 
+      (fun acc el -> 
+        transl_block None acc el ) 
+      base_block (List.tl blocks)
+  in
+  (match translated_blocks, global_counter with 
+  | Unguarded(body, _kind), Some gc ->
+    let arr = Ident.create_local "arr" in 
+    let len = Ident.create_local "len_var" in 
+    Llet(
+      Variable, Pintval, gc, int 0,
+      Llet(Strict, Pgenval, arr, body,
+        Llet(Alias, Pintval, len, 
+          Lprim((Parraylength(Paddrarray)), [Lvar(arr)], loc),
+          transl_concat_arrays (Lvar(arr)) (Lvar(len)) arrs array_kind gc)))
+  | Unguarded(body, _kind), None -> body 
+  | Guarded(body_and_len, _, id), Some gc -> 
+    let body = Ident.create_local "body" in
+    let len = Ident.create_local "len" in
+    let arr = Ident.create_local "arr" in 
+    Llet(
+      Variable, Pintval, gc, int 0,
+      Lstaticcatch(body_and_len, (id, [(body, Pgenval); (len, Pintval)]),
+        Llet(Strict, Pgenval, arr, Lvar(body),
```

Comment:
No need for this additional let, you can just use `Lvar(body)` instead of `Lvar(arr)` -- although I might use the name `arr` rather than `body` (and `arr_and_len`).